osu.Game/Skinning/Skin.cs

```diff
@@ -148,6 +149,28 @@ protected Skin(SkinInfo skin, IStorageResourceProvider? resources, IResourceStor
                     Logger.Error(ex, "Failed to load skin configuration.");
                 }
             }
+
+            if (layoutInfos.Values.Any())
+            {
+                int version = layoutInfos.Values.Min(l => l.Version);
+
+                for (int i = version + 1; i <= SkinLayoutInfo.LATEST_VERSION; i++)
+                    applyMigration(i);
```

Comment:
There should be no taking of `.Min()` here, and `applyMigration()` should be called in the inner loop over `layoutInfos.Values`.

Why? Presume you have two `LayoutInfo`s with different versions each. `LayoutInfo` X is on version 1, and `LayoutInfo` Y on version 2. Now when the latest version is bumped to 3, _Y will have the migration from version 1 to 2 called twice_. Which is _only fine_ if the migration is _idempotent_, as in it can _determine whether it has already run, and bail if it has_.

This is a _very not obvious_ property of a migration system; with these you generally assume that a migration is a _one-and-done_ affair. Thus I object to this type of handling, as an unnecessary complication. If targets are allowed to deviate in version, then migrations should be applied on a per-target basis.

Reply:
> and `applyMigration()` should be called in the inner loop over layoutInfos.Values.

Read OP.

The intention is that we should never have a case where there are two layouts with different migration versions, unless the user decided to mess around with the files. And, hopefully our migration system is good enough to not affect layouts which are already migrated even if the migration was ran again, but that's just a loose 
assumption.

That being said, with the concern raised both here and essentially https://github.com/ppy/osu/pull/30867#discussion_r1856466286, I will revert this and compare against the instantiation type instead. I realised at the end of the day if a configuration is null then the default components technically rely on the "instantiation type" (what type of skin class the skin is), so it's more logical to read the instantiation type and go from there instead.

Reply:
> The intention is that we should never have a case where there are two layouts with different migration versions, unless the user decided to mess around with the files.

You can't assume that people don't mess with the files, because they already do, and will become to do it more after external skin mounting is added.